bi/pkg/specs/wait.go

```diff
@@ -93,36 +101,35 @@ func jobFailed(conditions []batchv1.JobCondition) bool {
 	return false
 }
 
-func controlServerPodWatchOpts(ns string) *kube.WatchOptions {
+func controlServerDeployWatchOpts(ns string) *kube.WatchOptions {
 	return &kube.WatchOptions{
-		GVR:       schema.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"},
+		GVR:       schema.GroupVersionResource{Group: "apps", Version: "v1", Resource: "deployments"},
 		Namespace: ns,
-		ListOpts: metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(&metav1.LabelSelector{
-			MatchExpressions: []metav1.LabelSelectorRequirement{
-				{
-					Key:      "battery/app",
-					Operator: metav1.LabelSelectorOpIn,
-					Values:   []string{"battery-control-server"},
-				},
-			},
-		})},
-		Callback: func(u *unstructured.Unstructured) (bool, error) {
-			var pod corev1.Pod
-			err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &pod)
-			if err != nil {
-				slog.Debug(
-					"failed to convert unstructured object into typed resource",
-					slog.String("namespace", u.GetNamespace()),
-					slog.String("name", u.GetName()),
-					slog.Any("err", err),
-				)
-				return false, nil
-			}
+		ListOpts:  controlServerListOpts(),
+		Callback:  buildCallback(deployReady),
+	}
+}
 
-			// keep watching as long as the pod isn't running
-			return pod.Status.Phase == corev1.PodRunning, nil
+func controlServerListOpts() metav1.ListOptions {
+	return metav1.ListOptions{LabelSelector: metav1.FormatLabelSelector(&metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      "battery/app",
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{"battery-control-server"},
+			},
 		},
-	}
+	})}
+}
+
+func deployReady(deploy *appsv1.Deployment) (bool, error) {
+	status := deploy.Status
+
+	return status.ObservedGeneration > 1 &&
+			status.AvailableReplicas > 0 &&
+			status.UpdatedReplicas > 0 &&
+			status.ReadyReplicas > 0,
+		nil
```

Comment:
Change to watch the controlserver deployment instead of the pods. The `ObservedGeneration > 1` ensures that cs has started at least once and updated the deploy with the new secret key base, etc. The rest of the conditions make sure we have at least one ready pod on the newest version of the deploy.

Reply:
Please add that info as a comment somewhere